lib/trie/hash.go

```diff
@@ -5,195 +5,285 @@ package trie
 
 import (
 	"bytes"
-	"context"
+	"errors"
+	"fmt"
 	"hash"
 	"sync"
 
 	"github.com/ChainSafe/gossamer/lib/common"
 	"github.com/ChainSafe/gossamer/pkg/scale"
 	"golang.org/x/crypto/blake2b"
-	"golang.org/x/sync/errgroup"
 )
 
-// Hasher is a wrapper around a hash function
-type hasher struct {
-	hash     hash.Hash
-	tmp      bytes.Buffer
-	parallel bool // Whether to use parallel threads when hashing
+var encodingBufferPool = &sync.Pool{
+	New: func() interface{} {
+		const initialBufferCapacity = 1900000 // 1.9MB, from checking capacities at runtime
+		b := make([]byte, 0, initialBufferCapacity)
+		return bytes.NewBuffer(b)
+	},
 }
 
-// hasherPool creates a pool of Hasher.
-var hasherPool = sync.Pool{
+var digestBufferPool = &sync.Pool{
 	New: func() interface{} {
-		h, _ := blake2b.New256(nil)
-		var buf bytes.Buffer
-		// This allocation will be helpful for encoding keys. This is the min buffer size.
-		buf.Grow(700)
-
-		return &hasher{
-			tmp:  buf,
-			hash: h,
-		}
+		const bufferCapacity = 32
+		b := make([]byte, 0, bufferCapacity)
+		return bytes.NewBuffer(b)
 	},
 }
 
-// NewHasher create new Hasher instance
-func newHasher(parallel bool) *hasher {
-	h := hasherPool.Get().(*hasher)
-	h.parallel = parallel
-	return h
+var hasherPool = &sync.Pool{
+	New: func() interface{} {
+		hasher, err := blake2b.New256(nil)
+		if err != nil {
+			panic("cannot create Blake2b-256 hasher: " + err.Error())
+		}
+		return hasher
+	},
 }
 
-func (h *hasher) returnToPool() {
-	h.tmp.Reset()
-	h.hash.Reset()
-	hasherPool.Put(h)
-}
+func hashNode(n node, digestBuffer *bytes.Buffer) (err error) {
+	encodingBuffer := encodingBufferPool.Get().(*bytes.Buffer)
+	encodingBuffer.Reset()
+	defer encodingBufferPool.Put(encodingBuffer)
+
+	const parallel = false
```

Comment:
is this always false now? can we set it to not false by default or no?

Reply:
I have no idea. It should work the same, but I left it as is it was for now.

Memory stack wise, I would tend to think that would end up running 100(00)s of goroutines, leading to hundreds of MBs on the stack. I would rather have it run slower for now. We could have it in the future with `cpu cores` worker goroutines instead so it would be fast and with a low memory footprint, but definitely not in this PR :wink:

Reply:
I'd remove `parallel` completely then.  Not sure why we have this as an option if the caller of this method can never modify it.

Reply:
I kinda liked the parallel option (I think it increased the sync time, I did notice it is a bit slower on here, but would need to profile it to really make sure) but yeah either remove it or re-add it as an option I guess

Reply:
Let's leave it for now, to keep it similar to how it was in this PR. I'll do another PR to refactor it to use N workers with N = cpu cores. That would not use more memory and would be faster.